src/mixins/itext_behavior.mixin.js

```diff
@@ -524,24 +524,40 @@
     /**
      * Inserts a character where cursor is (replacing selection if one exists)
      * @param {String} _chars Characters to insert
+     * @param {Boolean} useCopiedStyle use fabric.copiedStyle
+     * @param {Boolean} skipUpdate trigger rendering and updates at the end of text insert
+     * @param {Object} styleObject Style to be inserted for the new char
      */
-    insertChars: function(_chars, useCopiedStyle) {
+    insertChars: function(_chars, useCopiedStyle, skipUpdate, styleObject) {
       if (this.selectionEnd - this.selectionStart > 1) {
         this._removeCharsFromTo(this.selectionStart, this.selectionEnd);
         this.setSelectionEnd(this.selectionStart);
       }
-      var isEndOfLine = this.text[this.selectionStart] === '\n';
-      this.text = this.text.slice(0, this.selectionStart) +
-        _chars + this.text.slice(this.selectionEnd);
-      this.insertStyleObjects(_chars, isEndOfLine, useCopiedStyle);
-      this.setSelectionStart(this.selectionStart + _chars.length);
-      this.setSelectionEnd(this.selectionStart);
-
-      this.canvas && this.canvas.renderAll();
-
-      this.setCoords();
-      this.fire('changed');
-      this.canvas && this.canvas.fire('text:changed', { target: this });
+      if (_chars.length > 1) {
+        var style;
+        for (var i = 0, len = _chars.length; i < len; i++) {
+          if (useCopiedStyle) {
+            style = fabric.copiedStyle[i];
+          }
+          this.insertChars(_chars[i], useCopiedStyle, i < len - 1, style);
+        }
+      }
+      else {
+        var isEndOfLine = this.text[this.selectionStart] === '\n';
```

Comment:
It seems we should move this whole block out of here into a separate method. Perhaps, `insertChar`?